common/unsafe/src/main/java/org/apache/spark/unsafe/UTF8StringBuilder.java

```diff
@@ -93,7 +93,38 @@ public void appendBytes(Object base, long offset, int length) {
     cursor += length;
   }
 
+  public void appendByte(byte singleByte) {
```

Comment:
Does this need to be a public method? Looks like we only use it below.

In a different direction: below we grow the buffer with either 1, 2, 3, or 4 bytes. In each case we call the `grow` method 1, 2, 3, or 4 times respectively. Does it make sense to have four different overloads each taking as inputs 1, 2, 3, or 4 bytes and do the growing once with the correct value?

Reply:
I put it as a public method in case anybody needs to append a single byte without having to allocate any objects on the heap (just as I would have to in the 'appendCodePoint' method if i decided to use the already existing 'appendBytes' method). I suppose that it is unnecessary to expose it with the assumption that somebody will need it at some point, as it can be exposed later if needed. I will change it to private.

For the second part, if i understood correctly, you proposed to make four new methods, where they respectively take 1,2,3,4 bytes that we want to append as parameters?
If thats the case, that would be much more efficient with respect to number of method calls (it would be 2 calls always), than the current implementation (which has 2n calls, where n is the number of appended bytes). In the current implementation, the number of buffer reallocations(due to calling 'grow') will be usually at most 1 per call of 'appendCodePoint', whereas in your proposed implementation it would be always at most 1. I had this in mind but I traded efficiency for code conciseness in this place. I think your proposal sounds good since my tradeoff was bad. Will implement it.

Reply:
> I put it as a public method in case anybody needs to append a single byte without having to allocate any objects on the heap

I agree with this, but we don't use it as such right now. I prefer to have it private and make it public as needed.

Reply:
> For the second part, if i understood correctly, you proposed to make four new methods, where they respectively take 1,2,3,4 bytes that we want to append as parameters?

Yes, this was the proposal. A completely different alternative it to discard the `appendByte` method altogether and put the actual code inline inside `appendCodePoint`.

Reply:
As @mkaravel and I discussed on slack, we decided to go with inlining all the byte appends into the the 'appendCodePoint' method, since there is less overhead and confusion regarding the 'appendByte' method.